### PR TITLE
add logout functionnality to logout button

### PR DESCRIPTION
**Pull request by marinovdeyan**
- add SharedPreferencesManager to avoid having Context being passed around everywhere (especially in UserViewModel)
- convert WorkerScheduler object to class and make it store the Context passed by MainActivity add function to unschedule worker
- adapted UserRepositoryFirestoreTest to these new modifications
- UserRepositoryFirestore now takes as arguments a WorkerScheduler and SharedPreferencesManager, to avoid having to store the Context directly in the URF.

### CODE REVIEW

**Review by daliaghosn — CHANGES_REQUESTED** [review]
**Positive Feedback:**
- Modular Design: The creation of SharedPreferencesManager encapsulates shared preferences operations, promoting modularity and reducing code duplication.
- Improved Code Maintainability: By managing shared preferences in a dedicated class, the code becomes easier to maintain and extend.
- Context Management: Avoiding the need to pass Context around, especially in UserViewModel, enhances the code's clarity and reduces potential errors.

**Points of Improvement:**
- Error Handling: Consider adding error handling mechanisms to manage potential issues when accessing shared preferences.
- Documentation: Adding comments or documentation for the methods within SharedPreferencesManager would improve code readability.

File: app/src/main/java/com/swent/suddenbump/worker/WorkerScheduler.kt [hunk 1]
```diff
@@ -35,11 +30,13 @@ object WorkerScheduler {
...
             .build()
 
-    // Enqueue the work request with a unique name to avoid duplication
-    WorkManager.getInstance(context)
+    WorkManager.getInstance(context.applicationContext)
         .enqueueUniquePeriodicWork(
             "LocationUpdateWorker", ExistingPeriodicWorkPolicy.REPLACE, workRequest)
+  }
 
-    Log.d("WorkerSuddenBump", "LocationUpdateWorker scheduled")
```
Comment (DanR2003) [inline app/src/main/java/com/swent/suddenbump/worker/WorkerScheduler.kt:43]: Good job for eliminating the no longer necessary logs!

File: app/src/main/java/com/swent/suddenbump/model/user/UserRepositoryFirestore.kt [hunk 2]
```diff
@@ -1100,18 +1097,15 @@ class UserRepositoryFirestore(private val db: FirebaseFirestore, private val con
...
    * Logs out the current user by updating shared preferences to remove login status and user ID.
    */
   override fun logoutUser() {
-    with(sharedPreferences.edit()) {
-      putBoolean("isLoggedIn", false)
-      putString("userId", null)
-      apply()
-    }
+    workerScheduler.unscheduleLocationUpdateWorker()
+    sharedPreferencesManager.clearPreferences()
```
Comment (DanR2003) [inline app/src/main/java/com/swent/suddenbump/model/user/UserRepositoryFirestore.kt:1108]: Very good implementation here: it's important to both unschedule the LocationUpdateWorker and clear the user preferences upon logging out.

**Review by DanR2003 — CHANGES_REQUESTED** [review]
- The logout functionality works well and the toast correctly informs the user about the logout being completed successfully, as it can be seen in the above screenshot:
Screenshot 2024-12-12 at 14 18 04

- On the emulator, when I try to login after logout, sometimes the app remains stuck inside the login screen. However, I believe this is a bug related to the emulator and not to the implementation because I haven't encountered it on my android phone, where the functionality worked well each time I tried.

- Documentation for the function defined in SharedPreferencesManager would be useful and would make reading the code easier, so consider adding it.

- Converting the WorkerScheduler object to  a class and making it store the Context is a very good decision, improving the clarity and modularity of the code, by eliminating the need to pass the Context every time

Overall this is a good PR focused on cleaning the UserRepositoryFirestore and making the necessary changes and additions to implement the ogOut functionality. After you add the requested documentation, this PR will be good to be merged into main in my opinion.

File: app/src/main/java/com/swent/suddenbump/worker/WorkerScheduler.kt
Comment (ifaresse) [inline app/src/main/java/com/swent/suddenbump/worker/WorkerScheduler.kt:1]: Why did you remove the documentation  in this file ?

**Review by ifaresse — CHANGES_REQUESTED** [review]
It's good that you wrote SharePreferencesManager to factor the sharedPreference linked function but maybe you could go further and regroup them in a new file "UserRepositorySharedPreference" as these function are not linked to firestore, this is only a suggestion. Apart from this restore the documentation you removed please, if it's ok with you.

File: app/src/main/java/com/swent/suddenbump/worker/WorkerScheduler.kt
Comment (marinovdeyan) [reply app/src/main/java/com/swent/suddenbump/worker/WorkerScheduler.kt:1]: Good question... I added it back!

**Comment by DanR2003** [comment]
> @daliaghosn Thank you for your feedback! I have added the documentation requested. Regarding the error handling, I am not sure exactly what is expected as catching the errors would require handling them every time we use the SharedPreferences. For instance, if we use some int or some boolean to check if the access to the SharedPreferences failed of succeeded, the functions won't be void anymore and we will have to match the error cases every single time we use them. Since the SharedPreferences are absolutely essential to the functionnality of our app, I am not exactly sure what kind of error handling you are expecting to be implemented. Could you expand on this a little or give me an example?
> 
> @ifaresse Thanks for the review, I have added back the documentation I removed by accident.
> 
> @DanR2003 Thanks for the review. The documentation requested has been added. I believe that the bug you are facing is not coming from the code, because I am unable to reproduce it both on the emulator and my Android phone. Did you have anything in the logs?
> 
> Anyways, if you guys could have a look at the modifications and tell me what you think, that would be amazing.

No, I did not have anything in the logs. But don't worry because I also tested on my phone and it worked every time so this was almost certainly just my emulator bugging

File: app/src/main/java/com/swent/suddenbump/model/user/SharedPreferencesManager.kt [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+   * @param key The key under which the value is stored.
+   * @param defaultValue The default value to return if the key does not exist.
+   * @return The boolean value associated with the key, or the default value if the key does not
+   *   exist.
+   */
+  fun getBoolean(key: String, defaultValue: Boolean = false): Boolean {
+    return sharedPreferences.getBoolean(key, defaultValue)
+  }
+
+  /** Clears all values stored in the shared preferences. */
```
Comment (DanR2003) [inline app/src/main/java/com/swent/suddenbump/model/user/SharedPreferencesManager.kt:57]: Very good. Nice to see you added the requested documentation :)

**Review by DanR2003 — APPROVED** [review]
I looked at the changed you added and everything looks good to me.

**Review by ifaresse — APPROVED** [review]
LGTM just be sure to merge the main in before, merging this PR

**Review by daliaghosn — APPROVED** [review]
You have implemented the required changes.

**Pull request merged**